### PR TITLE
new master

### PR DESCRIPTION
**Pull request by tranma**
quite a bit of savagery in here, psv and zebra will need to be detangled later

! @jystic @amosr

### CODE REVIEW

File: data/sea/51-piano.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#include "50-chord.h"
+
+typedef int error_t;
```
Comment (amosr) [inline data/sea/51-piano.h]:
```
typedef void error_t
```

File: data/sea/52-psv.h [hunk 2]
```diff
@@ -71,7 +71,21 @@ static ifleet_t * INLINE psv_alloc_fleet (iint_t max_chord_count);
...
+  , ifleet_t *fleet
+  );
+#else
+static ierror_loc_t INLINE psv_configure_fleet
+  ( const char *entity
+  , size_t entity_size
+  , const ichord_t **chord
+  , ifleet_t *fleet
+  );
+#endif
```
Comment (jystic) [inline data/sea/52-psv.h]: Unless it's easier to do it this way for now, we should be able to just delete all the `ichord_t` stuff, piano is designed to handle the ice chord use case as well

File: icicle-compiler/main/icicle-bench.hs [hunk 3]
```diff
@@ -43,18 +44,26 @@ pCommand = Command
...
+     = strOption $ long "output"
    pC
-     = argument str $ metavar "C"
+     = strOption $ long "code" <> short 'c'
+   pMode
+     = flip option (long "mode" <> value FlagSnapshot)
+     $ readerAsk >>= \case
+         "snapshot" -> return FlagSnapshot
+         "chords" -> return FlagChords
+         _ -> readerError "snapshot or chords"
```
Comment (jystic) [inline icicle-compiler/main/icicle-bench.hs]: s/chords/chord

File: icicle-compiler/src/Icicle/Sea/Eval/Base.hs [hunk 4]
```diff
@@ -255,28 +264,52 @@ seaCompile' options cache input programs = do
...
+                  Left e ->
+                    Savage.error $ "error merge: " <> show e
+                  Right _ ->
+                    return ()
+
+      return $ \ptr -> case chords of
+        Nothing -> do
+          -- FIXME never used in snapshot case, maybe wrap in ifdefs
+          n <- newForeignPtr_ nullPtr
+          withCPiano (ForeignPiano n) (go ptr)
```
Comment (jystic) [inline icicle-compiler/src/Icicle/Sea/Eval/Base.hs]: can maybe just call `go nullPtr`

**Comment by tranma** [comment]
once this builds, it should be able to run zebra chord

File: icicle-compiler/master.toml [hunk 5]
```diff
@@ -16,3 +16,7 @@
   ICICLE_CC_OPTIONS = "-Wall -Werror"
   HADDOCK = "true"
   HADDOCK_S3 = "$AMBIATA_HADDOCK_BRANCHES"
+  PUBLISH = "true"
+  PUBLISH_S3 = "$AMBIATA_ARTEFACTS_BRANCHES"
+  PUBLISH_EXECUTABLES = "icicle-bench"
+  TEST = "false"
```
Comment (jystic) [inline icicle-compiler/master.toml]: 👍

**Comment by tranma** [comment]
```
  connectivePrim p = case p of
    Min.PrimRelationGt -> primAnd
    Min.PrimRelationGe -> primAnd
    Min.PrimRelationLt -> primAnd
    Min.PrimRelationLe -> primAnd
    Min.PrimRelationEq -> primAnd
    Min.PrimRelationNe -> primOr
```

Ne is wrong... consider melting (x /= Some 0)

**Comment by amosr** [comment]
👍

**Comment by tranma** [comment]
"I got 90% through and it all looked good" -- Amos

**Pull request merged**